### ISSUE TITLE
Add Density option to the fog parameter

### ISSUE THREAD

**Issue by samuncle** [godotengine/godot#3806]
Hi

It would have been nice to have a density parameter for the fog. Currently we have falloff and color but density would allow to use fog to add a nice depth effect without occluding completely the world with the fog color.

I made an hard coded test for my own purpose, but of course that would be better to have a slider in the fog parameters in the world environment option.

```

diff --git a/drivers/gles2/shaders/material.glsl b/drivers/gles2/shaders/material.glsl
index ccd80bf..3442569 100644
--- a/drivers/gles2/shaders/material.glsl
+++ b/drivers/gles2/shaders/material.glsl
@@ -1205,7 +1205,7 @@ LIGHT_SHADER_CODE

 #ifdef USE_FOG

-               diffuse.rgb = mix(diffuse.rgb,fog_interp.rgb,fog_interp.a);
+               diffuse.rgb = mix(diffuse.rgb,fog_interp.rgb,fog_interp.a * 0.5);

 # if defined(LIGHT_TYPE_OMNI) || defined (LIGHT_TYPE_SPOT)
                diffuse.rgb = mix(mix(vec3(0.0),diffuse.rgb,attenuation),diffuse.rgb,const_light_mult);

```

**Comment by samuelgrigolato** [comment]
I gave this a little try, but I'm having some problems. The parameter in the editor works just fine, but when I try to use the value inside the shader, I get a white screen whenever I turn on the fog:

![fog_disabled](https://cloud.githubusercontent.com/assets/352108/16365062/b34df8e4-3bcb-11e6-8305-b0a3545f21d2.png)

![fog_enabled](https://cloud.githubusercontent.com/assets/352108/16365064/b83e1cc6-3bcb-11e6-9266-10daecfb35ca.png)

If I replace, in the shader, `fog_params.a` for `0.5` (or any other constant value) it works just fine. Could anyone point me where the problem may be? I tried also adding another uniform (`uniform highp float fog_density`) but had the same symptom. See my referenced commit for details.

**Comment by akien-mga** [comment]
Not critical for the upcoming 2.1, so moving to the next milestone.

**Comment by akien-mga** [comment]
Can someone confirm if this is implemented in the master branch?

**Comment by Calinou** [comment]
> Can someone confirm if this is implemented in the master branch?

No, the fog color's alpha value is ignored – if it was taken into account, it would allow decreasing the fog density.

**Comment by reduz** [comment]
maybe for 3.1

**Comment by akien-mga** [comment]
Fixed by #23248.
